include/tudocomp/coders/AdaptiveHuffmanCoder.hpp

```diff
@@ -0,0 +1,857 @@
+                gen_standard_code(final_code, symbol);
+            }
+
+            bool sign;
+            while (!final_code.empty()) {
+                sign = final_code.front();
+                final_code.pop();
+                os.write_bit(sign);
+            }
+
+            update(enc_codingtree, symbol);
+        }
+
+        /**
+         * Decode
+         */
+
+        // Pulls correct number of bits form input stream
+        inline uliteral_t get_char_from_code(tdc::io::BitIStream &is) {
+            uliteral_t temp;
+            temp &= 0x00;
+
+            unsigned int symbol_size = sizeof(uliteral_t)*CHAR_BIT;
+            for (unsigned int i = 0; i < symbol_size; i++) {
+                if (is.read_bit() == 1) {
+                    temp ^= 0x01;
+                }
+                if (i != 7) {
+                    temp <<= 1;
+                }
+            }
+
+            return temp;
+
+        }
+
+
+        /**
+         * \brief Decodes the next character from the input stream
+         * @param is
+         * @param dec_codingtree
+         * @return
+         */
+        inline uliteral_t decode(
+                tdc::io::BitIStream &is,
+                codingtree *dec_codingtree) {
+
+
+            node *temp = dec_codingtree->root;
+
+            std::queue<char> code_read;
+
+            // The tree is empty
+            if (temp == nullptr)
+            {
+                uliteral_t symbol = get_char_from_code(is);
+                update(dec_codingtree, symbol);
+                return symbol;
+            } else
+                // The tree exists, we need to walk it down until we reach a leaf
+            {
+                while (!temp->is_leaf()) {
+                    if (is.read_bit() == 0) {
+                        temp = temp->left;
+                        code_read.push('0');
+                    } else {
+                        temp = temp->right;
+                        code_read.push('1');
+                    }
+                }
+
+                if (temp->weight == 0)
+                    // The nyt-node is found
+                {
+                    uliteral_t symbol = get_char_from_code(is);
+                    update(dec_codingtree, symbol);
+                    return symbol;
+                } else
+                    // The corresponding node was found
+                {
+                    update(dec_codingtree, temp->symbol);
+                    return temp->symbol;
+                }
+            }
+        }
+
+    }//ns
+
+
+    class AdaptiveHuffmanCoder : public Algorithm {
+    public:
+        inline static Meta meta() {
+            Meta m("coder", "adaphuff", "Adaptive Huffman Coder");
+            return m;
+        }
+
+        inline static void initialize_codingtree(adaphuff::codingtree &codingtree) {
+            codingtree.root = nullptr;
+            codingtree.nyt = nullptr;
+            std::vector<bool> *dictionary = new std::vector<bool>(SYMBOL, false);
```

Comment:
In general its not necessary to put std data structures like a vector behind a new allocated pointer, since they all implement move semantic, and thus can be moved around cheaply without copies of the data.